modules/pbccs/main.nf

```diff
@@ -0,0 +1,49 @@
+// Import generic module functions
+include { initOptions; saveFiles; getSoftwareName } from './functions'
+
+params.options = [:]
+options        = initOptions(params.options)
+
+process PBCCS {
+    tag "$meta.id"
+    label 'process_low'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:meta, publish_by_meta:['id']) }
+
+    conda (params.enable_conda ? "bioconda::pbccs=6.0.0" : null)
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/pbccs:6.0.0--h9ee0642_2"
+    } else {
+        container "quay.io/biocontainers/pbccs:6.0.0--h9ee0642_2"
+    }
+
+    input:
+    tuple val(meta), path(bam), path(pbi), val(chunk_num), val(chunk_on)
```

Comment:
I'm not familiar with the tool nor the data, so sorry for the naïve question: but is the `chunk_{num,on}` a specific bit of metadata of the given BAM file? Or are they options? If the former, where is the `chunk` value defined - an upstream module?

`pbi` is OK to be with `bam` as I assume that's an [index file directly related to the BAM](https://github.com/nf-core/modules#inputoutput-options)?

Reply:
The `chunk_{num,on}` are options of `ccs`.
They have a parallelization purpose. 
Instead of splitting the bam file and run `ccs` on each part, you run `ccs` X times on the same bam file but you define the part to be processed.
`chunk_on` defines the number of parts of the bam file to process, and `chunk_num` select which part of the file is processed. 

In the pipeline I'm preparing, there is a `chunk` parameter. It creates a channel with all couples of `chunk_num` / `chunk_on` (1/100, 2/100, 3/100 ...) and then combine it to the `meta + bam + pbi` channel. 

I'm aware that program options should be passed through `addParams`at the module import. But because `chunk_num` is variable, I can't set it at module import.

The `pbi` file is the `bam` PacBio Index. `ccs` needs the pair of files for running.

Reply:
Ok. I so I guess then `chunk_new` would be dynamic, and would be associated with a specific BAM file.

However would `chunk_on` be typically the same across all BAM files? So in a workflow could you theoretically specify that _all_ BAM files would be split to the same number of chunks?

I ask as AFAIK (please correct me if I'm wrong @drpatelh ) we try to keep things as generic as possible and not make input channels for modules structured to a way specific to a certain pipeline (if that makes sense). Therefore we _prefer_ having `one input = one channel`

Reply:
Yes.
The idea is, for each (bam + pbi) input, ccs is run `chunk` times. 
So, if there's two bam files (samp1.bam + samp1.bam.pbi) and `chunk == 3`, the input channel will be: 
[samp1.bam, samp1.bam.pbi, 1, 3], 
[samp1.bam, samp1.bam.pbi, 2, 3], 
[samp1.bam, samp1.bam.pbi, 3, 3], 
[samp2.bam, samp2.bam.pbi, 1, 3], 
[samp2.bam, samp2.bam.pbi, 2, 3], 
[samp2.bam, samp2.bam.pbi, 3, 3]
Indeed, the `chunk_on` parameter is the same and could be defined at module import in addParams. 
I can the code update for that. 

I understand the point about `one input = one channel`. 
I can update the code to follow that.

Reply:
@jfy133 After code reviews, the `chunk_on` parameter cannot be passed at import with addParams. 
Indeed, the arg of the option `--chunk`  must be formatted like this: '`chunk_number/number_of_chunks`'. So the complete argument is variable. 
The best I can do is to reduce the number of arguments and merged `chunk_num` and `chunk_on` into one input (`chunk`)

Reply:
Ah ok, don't worry - I meant precisely that to have the chunks options as input channels, not as addParams. You have done what I meant anyway :)